File: python/print_network_speed.py
# 打印当前网速，适用于linux和windows
import os
import time
import psutil


def get_network_speed():
    # 获取当前网络传输数据量（字节）
    bytes_sent = psutil.net_io_counters().bytes_sent
    bytes_recv = psutil.net_io_counters().bytes_recv

    # 等待1秒钟
    time.sleep(1)

    # 再次获取网络传输数据量（字节）
    new_bytes_sent = psutil.net_io_counters().bytes_sent
    new_bytes_recv = psutil.net_io_counters().bytes_recv

    # 计算每秒发送/接收的字节数
    sent_per_sec = (new_bytes_sent - bytes_sent) / 1024
    recv_per_sec = (new_bytes_recv - bytes_recv) / 1024

    os.system('cls' if os.name == 'nt' else 'clear')
    if sent_per_sec < 1024:
        sent_per_sec_str = "\033[93m\u2191\033[0m: {:.2f} \033[93mKB/s\033[0m".format(sent_per_sec)
    else:
        sent_per_sec_str = "\033[93m\u2191\033[0m: {:.2f} \033[92mMB/s\033[0m".format(sent_per_sec/1024)
    if recv_per_sec < 1024:
        recv_per_sec_str = "\033[92m\u2193\033[0m: {:.2f} \033[93mKB/s\033[0m".format(recv_per_sec)
    else:
        recv_per_sec_str = "\033[92m\u2193\033[0m: {:.2f} \033[92mMB/s\033[0m".format(recv_per_sec/1024)
    print(sent_per_sec_str, recv_per_sec_str)
    print()


if __name__ == "__main__":
    while True:
        # 调用函数开始显示实时网速
        get_network_speed()

File: cpp/print_network_speed.cpp
# 实时打印当前网速，适用于linux
#include <iostream>
#include <fstream>
#include <string>
#include <chrono>
#include <thread>
#include <iomanip>

#ifdef _WIN32
    #define CLEAR_COMMAND "cls"
#else
    #define CLEAR_COMMAND "clear"
#endif

void clearScreen() {
    // 使用系统命令清屏
    system(CLEAR_COMMAND);
}

class NetworkSpeedMonitor {
public:
    NetworkSpeedMonitor(const std::string& interfaceName) : interfaceName_(interfaceName) {}

    void printCurrentSpeed() {
        std::cout << std::fixed << std::setprecision(2);

        unsigned long long rxBytesPrev = getRxBytes();
        unsigned long long txBytesPrev = getTxBytes();

        std::this_thread::sleep_for(std::chrono::seconds(1)); // 等待一秒

        unsigned long long rxBytesCurr = getRxBytes();
        unsigned long long txBytesCurr = getTxBytes();

        unsigned long long rxSpeed = rxBytesCurr - rxBytesPrev;
        unsigned long long txSpeed = txBytesCurr - txBytesPrev;

        clearScreen();

        std::cout << "Interface: " << "\033[34m" << interfaceName_ << "\033[0m" << "\t";

        // download
        if(rxSpeed < 1024)
        {
            std::cout << u8"\033[32m\u2193\033[0m " << rxSpeed << " \033[33mbytes/s\033[0m\t";
        }
        else
        {
            float rxSpeed_f = rxSpeed /1024.0;
            if(rxSpeed_f < 1024)
            {
                std::cout << u8"\033[32m\u2193\033[0m " << rxSpeed_f << " \033[32mKB/s\033[0m\t";
            }
            else
            {
                std::cout << u8"\033[32m\u2193\033[0m " << rxSpeed_f / 1024 << " \033[31mMB/s\033[0m\t";
            }
        }
        
        // upload
        if(txSpeed < 1024)
        {
            std::cout << u8"\033[33m\u2191\033[0m " << txSpeed << " \033[33mbytes/s\033[0m" << std::endl;
        }
        else
        {
            float txSpeed_f = txSpeed /1024.0;
            if(txSpeed_f < 1024)
            {
                std::cout << u8"\033[33m\u2191\033[0m " << txSpeed_f << " \033[32mKB/s\033[0m" << std::endl;
            }
            else
            {
                std::cout << u8"\033[33m\u2191\033[0m " << txSpeed_f / 1024 << " \033[32mMB/s\033[0m" << std::endl;
            }
        }
        
    }

private:
    unsigned long long getRxBytes() const {
        return getInterfaceBytes("rx_bytes");
    }

    unsigned long long getTxBytes() const {
        return getInterfaceBytes("tx_bytes");
    }

    unsigned long long getInterfaceBytes(const std::string& type) const {
        std::ifstream file("/sys/class/net/" + interfaceName_ + "/statistics/" + type);
        if (!file.is_open()) {
            std::cerr << "\033[31mError opening file for interface: " << interfaceName_ << "\033[0m" << ".\nCheck NIC name with \"ifconfig\"." << std::endl;
            return 0;
        }

        unsigned long long bytes;
        file >> bytes;
        file.close();

        return bytes;
    }

private:
    std::string interfaceName_;
};

int main() {
    NetworkSpeedMonitor monitor("ens33");

    // 打印当前网速
    while(true)
    {
        monitor.printCurrentSpeed();
    }
    return 0;
}
